packages/flutter_tools/lib/src/macos/macos_device.dart

```diff
@@ -2,145 +2,67 @@
-    DebuggingOptions debuggingOptions,
-    Map<String, dynamic> platformArgs,
-    bool prebuiltApplication = false,
-    bool ipv6 = false,
+    BuildInfo buildInfo,
   }) async {
-    if (!prebuiltApplication) {
-      Cache.releaseLockEarly();
-      await buildMacOS(
-        flutterProject: FlutterProject.current(),
-        buildInfo: debuggingOptions?.buildInfo,
-        targetOverride: mainPath,
-      );
-    }
-
-    // Ensure that the executable is locatable.
-    final String executable = package.executable(debuggingOptions?.buildInfo?.mode);
-    if (executable == null) {
-      printError('Unable to find executable to run');
-      return LaunchResult.failed();
-    }
-
-    _lastBuiltMode = debuggingOptions?.buildInfo?.mode;
-    final Process process = await processManager.start(<String>[
-      executable,
-    ]);
-    if (debuggingOptions?.buildInfo?.isRelease == true) {
-      return LaunchResult.succeeded();
-    }
-    _deviceLogReader.initializeProcess(process);
-    final ProtocolDiscovery observatoryDiscovery = ProtocolDiscovery.observatory(_deviceLogReader);
-    try {
-      final Uri observatoryUri = await observatoryDiscovery.uri;
-      // Bring app to foreground.
-      await processManager.run(<String>[
-        'open', package.applicationBundle(debuggingOptions?.buildInfo?.mode),
-      ]);
-      return LaunchResult.succeeded(observatoryUri: observatoryUri);
-    } catch (error) {
-      printError('Error waiting for a debug connection: $error');
-      return LaunchResult.failed();
-    } finally {
-      await observatoryDiscovery.cancel();
-    }
+    await buildMacOS(
+      flutterProject: FlutterProject.current(),
+      buildInfo: buildInfo,
+      targetOverride: mainPath,
+    );
   }
 
-  // TODO(jonahwilliams): implement using process manager.
-  // currently we rely on killing the isolate taking down the application.
   @override
-  Future<bool> stopApp(covariant MacOSApp app) async {
-    return killProcess(app.executable(_lastBuiltMode));
+  String executablePathForDevice(covariant MacOSApp package, BuildMode buildMode) {
+    return package.executable(buildMode);
   }
 
   @override
-  Future<TargetPlatform> get targetPlatform async => TargetPlatform.darwin_x64;
-
-  // Since the host and target devices are the same, no work needs to be done
-  // to uninstall the application.
-  @override
-  Future<bool> uninstallApp(ApplicationPackage app) async => true;
-
-  @override
-  bool isSupportedForProject(FlutterProject flutterProject) {
-    return flutterProject.macos.existsSync();
+  void onAttached(covariant MacOSApp package, BuildMode buildMode, Process process) {
+    // Bring app to foreground. Ideally this would be done post-launch rather
+    // than post-attach, since this won't run for release builds, but there's
+    // no general-purpose way of knowing when a process is far enoug along in
+    // the launch process for 'open' to foreground it.
+    processManager.run(<String>[
```

Comment:
Should this be awaited in some way?

Reply:
I can wire it up as async/await if you think it's clearer. I don't think it actually matters though (unless I'm forgetting something) since nothing in the tool cares whether it's foreground or background.

Reply:
Yeah, I don't think we necessarily need to block anything on this - but we might want to stick a `.then` to handle failures

Reply:
Added error logging in a `then`. Feel free to alter if you had something else in mind since I'm not sure when I'll next have a chance to iterate on it.